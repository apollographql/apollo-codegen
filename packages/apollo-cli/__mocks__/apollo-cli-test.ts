import Nock from "@fancy-test/nock";
import * as Test from "@oclif/test";
export { expect } from "@oclif/test";
import { mockConsole } from "heroku-cli-util";

export const test = Test.test.register("nock", Nock);
mockConsole();
