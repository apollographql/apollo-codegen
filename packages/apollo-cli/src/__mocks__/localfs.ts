import {Volume, createFsFromVolume} from 'memfs';
import {patchFs} from 'fs-monkey';

export const vol = Volume.fromJSON({});
export const fs = createFsFromVolume(vol);

export function withGlobalFS<T>(thunk: () => T): T {
  const unpatch = patchFs(vol);
  const ret = thunk();
  unpatch();
  return ret;
}
