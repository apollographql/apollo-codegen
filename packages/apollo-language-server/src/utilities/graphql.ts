import {
  GraphQLSchema,
  GraphQLCompositeType,
  GraphQLObjectType,
  GraphQLInterfaceType,
  GraphQLUnionType,
  GraphQLField,
  FieldNode,
  SchemaMetaFieldDef,
  TypeMetaFieldDef,
  TypeNameMetaFieldDef,
  ASTNode,
  Kind,
  NameNode,
  TypeSystemDefinitionNode,
  TypeSystemExtensionNode
} from "graphql";

export function isNode(maybeNode: any): maybeNode is ASTNode {
  return maybeNode && typeof maybeNode.kind === "string";
}

export type NamedNode = ASTNode & {
  name: NameNode;
};

export function isNamedNode(node: ASTNode): node is NamedNode {
  return "name" in node;
}

export function isTypeSystemDefinitionNode(
  node: ASTNode
): node is TypeSystemDefinitionNode {
  switch (node.kind) {
    case Kind.SCHEMA_DEFINITION:
    case Kind.SCALAR_TYPE_DEFINITION:
    case Kind.OBJECT_TYPE_DEFINITION:
    case Kind.INTERFACE_TYPE_DEFINITION:
    case Kind.UNION_TYPE_DEFINITION:
    case Kind.ENUM_TYPE_DEFINITION:
    case Kind.INPUT_OBJECT_TYPE_DEFINITION:
    case Kind.DIRECTIVE_DEFINITION:
      return true;
    default:
      return false;
  }
}

export function isTypeSystemExtensionNode(
  node: ASTNode
): node is TypeSystemExtensionNode {
  switch (node.kind) {
    // case Kind.SCHEMA_EXTENSION:
    case Kind.SCALAR_TYPE_EXTENSION:
    case Kind.OBJECT_TYPE_EXTENSION:
    case Kind.INTERFACE_TYPE_EXTENSION:
    case Kind.UNION_TYPE_EXTENSION:
    case Kind.ENUM_TYPE_EXTENSION:
    case Kind.INPUT_OBJECT_TYPE_EXTENSION:
      return true;
    default:
      return false;
  }
}

export function highlightNodeForNode(node: ASTNode): ASTNode {
  switch (node.kind) {
    case Kind.VARIABLE_DEFINITION:
      return node.variable;
    default:
      return isNamedNode(node) ? node.name : node;
  }
}

/**
 * Not exactly the same as the executor's definition of getFieldDef, in this
 * statically evaluated environment we do not always have an Object type,
 * and need to handle Interface and Union types.
 */
export function getFieldDef(
  schema: GraphQLSchema,
  parentType: GraphQLCompositeType,
  fieldAST: FieldNode
): GraphQLField<any, any> | undefined {
  const name = fieldAST.name.value;
  if (
    name === SchemaMetaFieldDef.name &&
    schema.getQueryType() === parentType
  ) {
    return SchemaMetaFieldDef;
  }
  if (name === TypeMetaFieldDef.name && schema.getQueryType() === parentType) {
    return TypeMetaFieldDef;
  }
  if (
    name === TypeNameMetaFieldDef.name &&
    (parentType instanceof GraphQLObjectType ||
      parentType instanceof GraphQLInterfaceType ||
      parentType instanceof GraphQLUnionType)
  ) {
    return TypeNameMetaFieldDef;
  }
  if (
    parentType instanceof GraphQLObjectType ||
    parentType instanceof GraphQLInterfaceType
  ) {
    return parentType.getFields()[name];
  }

  return undefined;
}
