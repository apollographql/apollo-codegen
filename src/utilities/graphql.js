// Utility functions extracted from graphql-js

import {
  SchemaMetaFieldDef,
  TypeMetaFieldDef,
  TypeNameMetaFieldDef,
  GraphQLObjectType,
  GraphQLInterfaceType,
  GraphQLUnionType
} from 'graphql';

/**
 * Extracts the root type of the operation from the schema.
 */
export function getOperationRootType(schema, operation) {
  switch (operation.operation) {
    case 'query':
      return schema.getQueryType();
    case 'mutation':
      const mutationType = schema.getMutationType();
      if (!mutationType) {
        throw new GraphQLError(
          'Schema is not configured for mutations',
          [operation]
        );
      }
      return mutationType;
    case 'subscription':
      const subscriptionType = schema.getSubscriptionType();
      if (!subscriptionType) {
        throw new GraphQLError(
          'Schema is not configured for subscriptions',
          [operation]
        );
      }
      return subscriptionType;
    default:
      throw new GraphQLError(
        'Can only compile queries, mutations and subscriptions',
        [operation]
      );
  }
}

/**
 * Not exactly the same as the executor's definition of getFieldDef, in this
 * statically evaluated environment we do not always have an Object type,
 * and need to handle Interface and Union types.
 */
export function getFieldDef(schema, parentType, fieldAST) {
  const name = fieldAST.name.value;
  if (name === SchemaMetaFieldDef.name &&
      schema.getQueryType() === parentType) {
    return SchemaMetaFieldDef;
  }
  if (name === TypeMetaFieldDef.name &&
      schema.getQueryType() === parentType) {
    return TypeMetaFieldDef;
  }
  if (name === TypeNameMetaFieldDef.name &&
      (parentType instanceof GraphQLObjectType ||
       parentType instanceof GraphQLInterfaceType ||
       parentType instanceof GraphQLUnionType)
  ) {
    return TypeNameMetaFieldDef;
  }
  if (parentType instanceof GraphQLObjectType ||
      parentType instanceof GraphQLInterfaceType) {
    return parentType.getFields()[name];
  }
}
