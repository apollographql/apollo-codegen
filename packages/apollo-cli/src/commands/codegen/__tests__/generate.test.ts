jest.mock("apollo-codegen-core/lib/localfs", () => {
  return require("../../../__mocks__/localfs");
});

// this is because of herkou-cli-utils hacky mocking system on their console logger
import { stdout, mockConsole } from "heroku-cli-util";
import * as path from "path";
import * as fs from "fs";
import { test as setup } from "apollo-cli-test";
import { introspectionQuery, print, execute, buildSchema } from "graphql";
import gql from "graphql-tag";
import { fs as mockFS, vol } from "apollo-codegen-core/lib/localfs";

const test = setup.do(() => mockConsole());
const fullSchema = execute(
  buildSchema(
    fs.readFileSync(
      path.resolve(__dirname, "../../schema/__tests__/fixtures/schema.graphql"),
      {
        encoding: "utf-8"
      }
    )
  ),
  gql(introspectionQuery)
).data;

const simpleQuery = fs.readFileSync(
  path.resolve(__dirname, "./fixtures/simpleQuery.graphql")
);

beforeEach(() => {
  vol.reset();
  vol.fromJSON({
    __blankFileSoDirectoryExists: ""
  });
});

jest.setTimeout(15000);

describe("successful codegen", () => {
  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "queryOne.graphql": simpleQuery.toString()
      });
    })
    .command(["codegen:generate", "--schema=schema.json", "API.swift"])
    .it("infers Swift target and writes types", () => {
      expect(mockFS.readFileSync("API.swift").toString()).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "queryOne.graphql": simpleQuery.toString()
      });
    })
    .command(["codegen:generate", "--schema=schema.json", "--operationIdsPath=myOperationIDs.json", "API.swift"])
    .it("generates operation IDs files when flag is set", () => {
      expect(mockFS.readFileSync("myOperationIDs.json").toString()).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "queryOne.graphql": simpleQuery.toString(),
        "queryTwo.graphql": otherQuery.toString(),
        "outDirectory/__create_this_directory": ""
      });
    })
    .command(["codegen:generate", "--schema=schema.json", "--only=queryTwo.graphql", "--target=swift", "outDirectory"])
    .it("handles only flag for Swift target", () => {
      expect(vol.toJSON("outDirectory")).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "queryOne.graphql": simpleQuery.toString()
      });
    })
    .command(["codegen:generate", "--schema=schema.json", "API.scala"])
    .it("infers Scala target and writes types", () => {
      expect(mockFS.readFileSync("API.scala").toString()).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "queryOne.graphql": simpleQuery.toString()
      });
    })
    .command(["codegen:generate", "--schema=schema.json", "--outputFlat", "API.ts"])
    .it("infers TypeScript target and writes types", () => {
      expect(mockFS.readFileSync("API.ts").toString()).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "queryOne.graphql": simpleQuery.toString()
      });
    })
    .command(["codegen:generate", "--schema=schema.json", "--outputFlat", "API.js"])
    .it("infers Flow target and writes types", () => {
      expect(mockFS.readFileSync("API.js").toString()).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "queryOne.graphql": simpleQuery.toString()
      });
    })
    .command(["codegen:generate", "--schema=schema.json", "--useFlowExactObjects", "API.js"])
    .it("writes exact Flow types when the flag is set", () => {
      expect(mockFS.readFileSync("API.js").toString()).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "queryOne.graphql": simpleQuery.toString()
      });
    })
    .command(["codegen:generate", "--schema=schema.json", "--useFlowReadOnlyTypes", "API.js"])
    .it("writes read-only Flow types when the flag is set", () => {
      expect(mockFS.readFileSync("API.js").toString()).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "queryOne.graphql": simpleQuery.toString()
      });
    })
    .command(["codegen:generate", "--schema=schema.json", "operations.json"])
    .it("infers JSON target and writes operations", () => {
      expect(
        mockFS.readFileSync("operations.json").toString()
      ).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "directory/component.tsx": `
          gql\`
            query SimpleQuery {
              hello
            }
          \`;
        `
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--queries=**/*.tsx",
      "--target=typescript"
    ])
    .it("writes TypeScript types into a __generated__ directory next to sources when no output is set", () => {
      expect(
        mockFS.readFileSync("directory/__generated__/SimpleQuery.ts").toString()
      ).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "directory/component.jsx": `
          gql\`
            query SimpleQuery {
              hello
            }
          \`;
        `
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--queries=**/*.jsx",
      "--target=flow"
    ])
    .it("writes Flow types into a __generated__ directory next to sources when no output is set", () => {
      expect(
        mockFS.readFileSync("directory/__generated__/SimpleQuery.js").toString()
      ).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "directory/component.tsx": `
          gql\`
            query SimpleQuery {
              hello
            }
          \`;
        `
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--queries=**/*.tsx",
      "--target=typescript",
      "__foo__"
    ])
    .it(
      "writes TypeScript types to a custom directory next to sources when output is set",
      () => {
        expect(
          mockFS
            .readFileSync("directory/__foo__/SimpleQuery.ts")
            .toString()
        ).toMatchSnapshot();
      }
    );

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "directory/component.jsx": `
          gql\`
            query SimpleQuery {
              hello
            }
          \`;
        `
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--queries=**/*.jsx",
      "--target=flow",
      "__foo__"
    ])
    .it(
      "writes Flow types to a custom directory next to sources when output is set",
      () => {
        expect(
          mockFS
            .readFileSync("directory/__foo__/SimpleQuery.js")
            .toString()
        ).toMatchSnapshot();
      }
    );

    test
      .do(() => {
        vol.fromJSON({
          "schema.json": JSON.stringify(fullSchema.__schema),
          "directory/component.tsx": `
            gql\`
              query SimpleQuery {
                hello
              }
            \`;
          `
        });
      })
      .command([
        "codegen:generate",
        "--schema=schema.json",
        "--queries=**/*.tsx",
        "--target=typescript",
        ""
      ])
      .it(
        "writes TypeScript types next to sources when output is set to empty string",
        () => {
          expect(
            mockFS
              .readFileSync("directory/SimpleQuery.ts")
              .toString()
          ).toMatchSnapshot();
        }
      );

    test
      .do(() => {
        vol.fromJSON({
          "schema.json": JSON.stringify(fullSchema.__schema),
          "directory/component.jsx": `
            gql\`
              query SimpleQuery {
                hello
              }
            \`;
          `
        });
      })
      .command([
        "codegen:generate",
        "--schema=schema.json",
        "--queries=**/*.jsx",
        "--target=flow",
        ""
      ])
      .it(
        "writes Flow types next to sources when output is set to empty string",
        () => {
          expect(
            mockFS
              .readFileSync("directory/SimpleQuery.js")
              .toString()
          ).toMatchSnapshot();
        }
      );
});

describe("error handling", () => {
  test
    .command(["codegen:generate", "--target=foobar"])
    .catch(err => expect(err.message).toMatch(/Unsupported target: foobar/))
    .it("errors with an unsupported target");

  test
    .command(["codegen:generate", "--target=swift"])
    .catch(err =>
      expect(err.message).toMatch(/The output path must be specified/)
    )
    .it("errors when no output file is provided");

  test
    .command(["codegen:generate", "output-file"])
    .catch(err => expect(err.message).toMatch(/Could not infer target from output file type, please use --target/))
    .it("errors when target cannot be inferred");
});
