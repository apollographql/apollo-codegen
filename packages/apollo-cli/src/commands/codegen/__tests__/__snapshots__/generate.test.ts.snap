// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`successful codegen generates operation IDs files when flag is set 1`] = `
"{
  \\"4de1e386589b0853a102a87a3f21ca25266116517e59c1e8be9442e2571f00bc\\": {
    \\"name\\": \\"SimpleQuery\\",
    \\"source\\": \\"query SimpleQuery {\\\\n  hello\\\\n}\\"
  }
}"
`;

exports[`successful codegen handles only flag for Swift target 1`] = `
Object {
  "/Users/shadaj/work/apollo-cli/packages/apollo-cli/outDirectory/Types.graphql.swift": "//  This file was automatically generated and should not be edited.

import Apollo",
  "/Users/shadaj/work/apollo-cli/packages/apollo-cli/outDirectory/__create_this_directory": "",
  "/Users/shadaj/work/apollo-cli/packages/apollo-cli/outDirectory/queryTwo.graphql.swift": "//  This file was automatically generated and should not be edited.

import Apollo

public final class OtherQueryQuery: GraphQLQuery {
  public let operationDefinition =
    \\"query OtherQuery {\\\\n  hello\\\\n}\\"

  public init() {
  }

  public struct Data: GraphQLSelectionSet {
    public static let possibleTypes = [\\"Query\\"]

    public static let selections: [GraphQLSelection] = [
      GraphQLField(\\"hello\\", type: .nonNull(.scalar(String.self))),
    ]

    public private(set) var resultMap: ResultMap

    public init(unsafeResultMap: ResultMap) {
      self.resultMap = unsafeResultMap
    }

    public init(hello: String) {
      self.init(unsafeResultMap: [\\"__typename\\": \\"Query\\", \\"hello\\": hello])
    }

    public var hello: String {
      get {
        return resultMap[\\"hello\\"]! as! String
      }
      set {
        resultMap.updateValue(newValue, forKey: \\"hello\\")
      }
    }
  }
}",
}
`;

exports[`successful codegen infers Flow target and writes types 1`] = `
"

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export type SimpleQuery = {
  hello: string
};

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//=============================================================="
`;

exports[`successful codegen infers JSON target and writes operations 1`] = `
"{
	\\"operations\\": [
		{
			\\"filePath\\": \\"queryOne.graphql\\",
			\\"operationName\\": \\"SimpleQuery\\",
			\\"operationType\\": \\"query\\",
			\\"rootType\\": \\"Query\\",
			\\"variables\\": [],
			\\"source\\": \\"query SimpleQuery {\\\\n  hello\\\\n}\\",
			\\"fields\\": [
				{
					\\"responseName\\": \\"hello\\",
					\\"fieldName\\": \\"hello\\",
					\\"type\\": \\"String!\\",
					\\"isConditional\\": false,
					\\"isDeprecated\\": false
				}
			],
			\\"fragmentSpreads\\": [],
			\\"inlineFragments\\": [],
			\\"fragmentsReferenced\\": [],
			\\"sourceWithFragments\\": \\"query SimpleQuery {\\\\n  hello\\\\n}\\",
			\\"operationId\\": \\"4de1e386589b0853a102a87a3f21ca25266116517e59c1e8be9442e2571f00bc\\"
		}
	],
	\\"fragments\\": [],
	\\"typesUsed\\": []
}"
`;

exports[`successful codegen infers Scala target and writes types 1`] = `
"//  This file was automatically generated and should not be edited.


object SimpleQueryQuery extends com.apollographql.scalajs.GraphQLQuery {
  val operationString =
    \\"query SimpleQuery {\\" +
    \\"  hello\\" +
    \\"}\\"
  val operation = com.apollographql.scalajs.gql(operationString)

  type Variables = Unit

  case class Data(hello: String) {
  }

  object Data {
    val possibleTypes = scala.collection.Set(\\"Query\\")
  }
}"
`;

exports[`successful codegen infers Swift target and writes types 1`] = `
"//  This file was automatically generated and should not be edited.

import Apollo

public final class SimpleQueryQuery: GraphQLQuery {
  public let operationDefinition =
    \\"query SimpleQuery {\\\\n  hello\\\\n}\\"

  public init() {
  }

  public struct Data: GraphQLSelectionSet {
    public static let possibleTypes = [\\"Query\\"]

    public static let selections: [GraphQLSelection] = [
      GraphQLField(\\"hello\\", type: .nonNull(.scalar(String.self))),
    ]

    public private(set) var resultMap: ResultMap

    public init(unsafeResultMap: ResultMap) {
      self.resultMap = unsafeResultMap
    }

    public init(hello: String) {
      self.init(unsafeResultMap: [\\"__typename\\": \\"Query\\", \\"hello\\": hello])
    }

    public var hello: String {
      get {
        return resultMap[\\"hello\\"]! as! String
      }
      set {
        resultMap.updateValue(newValue, forKey: \\"hello\\")
      }
    }
  }
}"
`;

exports[`successful codegen infers TypeScript target and writes types 1`] = `
"

/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export interface SimpleQuery {
  hello: string;
}

/* tslint:disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//=============================================================="
`;

exports[`successful codegen writes Flow types next to sources when no output is set 1`] = `
"

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export type SimpleQuery = {
  hello: string
};

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//=============================================================="
`;

exports[`successful codegen writes TypeScript types next to sources when no output is set 1`] = `
"

/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export interface SimpleQuery {
  hello: string;
}

/* tslint:disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//=============================================================="
`;

exports[`successful codegen writes exact Flow types when the flag is set 1`] = `
"

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export type SimpleQuery = {|
  hello: string
|};

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//=============================================================="
`;

exports[`successful codegen writes read-only Flow types when the flag is set 1`] = `
"

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export type SimpleQuery = {
  +hello: string
};

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//=============================================================="
`;
